vermouth/rcsu/get_go.py

```diff
@@ -0,0 +1,164 @@
+        self.res_dist = res_dist
+        self.itp_file = open('GoVirtIncludes.itp', "w", encoding='UTF-8')
+        self.prefix = prefix
+        self.res_graph = res_graph
+        self.water_bias = water_bias
+        self.idp_water_bias = idp_water_bias
+        self.regions = regions
+
+    def write_go_structural_bias(self, molecule):
+        chain_id_to_resnode = {}
+        # find all pairs of residues that are within bonded distance of
+        # self.res_dist
+        connected_pairs = dict(nx.all_pairs_shortest_path_length(self.res_graph,
+                                                            cutoff=self.res_dist))
+
+        # for each residue collect the chain and residue in a dict
+        # we use this later for identifying the residues from the
+        # contact map
+        for resnode in self.res_graph.nodes:
+            chain = self.res_graph.nodes[resnode].get('chain', None)
+            # in vermouth within a molecule all resid are unique
+            # when merging multiple chains we store the old resid
+            # the go model always references the input resid i.e.
+            # the _old_resid
+            resid = self.res_graph.nodes[resnode].get('_old_resid')
+            chain_id_to_resnode[(chain, resid)] = resnode
+
+        for contact in self.contact_map:
+            resIDA, chainA, resIDB, chainB = contact
+            # identify the contact in the residue graph based on
+            # chain ID and resid
+            resA = chain_id_to_resnode[(chainA, resIDA)]
+            resB = chain_id_to_resnode[(chainB, resIDB)]
+            # make sure that both residues are not connected
+            # note: contacts should be symmteric so we should be
+            # able to get rid of the second if clause
+            if resB not in connected_pairs[resA] and resA not in connected_pairs[resB]:
+                # now we lookup the backbone nodes within the residue contact
+                bb_node_A = _get_bb_nodes(molecule, self.res_graph.nodes[resA]['graph'].nodes)
+                bb_node_B = _get_bb_nodes(molecule, self.res_graph.nodes[resB]['graph'].nodes)
+                # compute the distance between bb-beads
+                dist = np.linalg.norm(molecule.nodes[bb_node_A]['position'] -
+                                      molecule.nodes[bb_node_B]['position'])
+                # verify that the distance between BB-beads satisfies the
+                # cut-off criteria
+                if dist > self.cutoff_short or dist < self.cutoff_long:
+                    # compute the LJ sigma paramter for this contact
+                    sigma = dist / 1.12246204830
+                    # find the go virtual-sites for this residue
+                    # probably can be done smarter but mehhhh
+                    nodeA = _get_go_type(molecule, resid=resIDA, chain=chainA, prefix=self.prefix)
+                    nodeB = _get_go_type(molecule, resid=resIDB, chain=chainB, prefix=self.prefix)
```

Comment:
The only problem I see is that you'll look these up multiple times per residue. We could cache/store/whatever them as resnode attributes, but profile first.

Reply:
You already gain if you do:
```suggestion
                    # find the go virtual-sites for this residue
                    # probably can be done smarter but mehhhh
                    nodeA = _get_go_type(self.res_graph.nodes[resA]['nodes'], resid=resIDA, chain=chainA, prefix=self.prefix)
                    nodeB = _get_go_type(self.res_graph.nodes[resB]['nodes'], resid=resIDB, chain=chainB, prefix=self.prefix)
```
(check spelling)

Reply:
well no because the go sides are not part of the residues in this stage, because of the issue pointed out earlier. But let's just make them part of the residue and screw downstream piplines